File: functional01-java/src/main/java/org/example/vavr/login/LoginController.java
package org.example.vavr.login;

import io.vavr.Tuple;
import io.vavr.Tuple2;
import io.vavr.control.Either;

public class LoginController {

    private UserDAO userDAO;

    public LoginController(UserDAO userDAO) {
        this.userDAO = userDAO;
    }

    /**
     *
     * @return Either
     *  left - String - error message
     *  right - String - confirmation
     */
    public Either<String, String> performLogin(String name, String password) {
        return validateName(Tuple.of(name, password))
                .flatMap(this::validatePassword)
                .flatMap(this::findUser)
                .flatMap(user -> checkPassword(user, password));
    }

    public Either<String, Tuple2<String, String>> validateName(Tuple2<String, String> t) {
        String name = t._1();
        if (name == null || name.equals(""))
            return Either.left("Invalid Name");

        return Either.right(t);
    }

    public Either<String, Tuple2<String, String>> validatePassword(Tuple2<String, String> t) {
        String password = t._2();
        if (password == null || password.equals(""))
            return Either.left("Invalid Password");

        return Either.right(t);
    }

    public Either<String, User> findUser(Tuple2<String, String> t) {
        var op = userDAO.getUsers().stream()
                .filter(user -> user.name().equals(t._1()))
                .findFirst();
        if (op.isPresent())
            return Either.right(op.get());

        return Either.left("User does not exist");
    }

    public Either<String, String> checkPassword(User user, String password) {
        if (!user.password().equals(password))
            return Either.left("Incorrect password");

        return Either.right(user.name() + " logged");
    }

}

File: functional-js/proj02/loginController/loginController.js
const { UserDAO } = require('./userDao');

const userDao = new UserDAO();

function performLogin(request) {
    return validateName(request)
        .then(validatePassword)
        .then(findUser)
        .then(user => checkPassword(user, request.password));
}

function findUser({ name }) {
    console.log('>>findUser');
    const user = userDao.getUsers().find(u => u.username == name);
    return (user != undefined) ? Promise.resolve(user) : Promise.reject("User does not exist");
}

function checkPassword(user, password) {
    console.log('>>checkPassword');

    if (user.password !== password)
        return Promise.reject("Incorrect password");

    return Promise.resolve(`${user.username} logged`)
}

function validateName(request) {
    console.log('>>validateName');

    const name = request.name;
    if (name == undefined || name == '')
        return Promise.reject("Invalid name");

    return Promise.resolve(request);
}

function validatePassword(request) {
    console.log('>>validatePassword');
    const password = request.password;
    if (password == undefined || password == '')
        return Promise.reject("Invalid password");

    return Promise.resolve(request);
}

performLogin({ name: "user4", password: "pass4" })
    .then(res => console.log(`Success: ${res}`))
    .catch(err => console.log(`Error: ${err}`));